polynote-server/src/main/scala/polynote/server/repository/NotebookRepository.scala

```diff
@@ -160,6 +166,34 @@ abstract class FileBasedRepository[F[_]](implicit F: ConcurrentEffect[F], contex
     }
   }
 
+  def renameNotebook(oldPath: String, newPath: String): F[String] = {
+    val ext = s".$defaultExtension"
+    val withExt = newPath.replaceFirst("""^/+""", "").stripSuffix(ext) + ext
+
+    if (relativeDepth(withExt) > maxDepth) {
+      F.raiseError(new IllegalArgumentException(s"Input path ($newPath) too deep, maxDepth is $maxDepth"))
+    } else {
+      (notebookExists(oldPath), notebookExists(withExt)).mapN(_ -> _).flatMap {
+        case (false, _)    => F.raiseError(new FileNotFoundException(s"File $oldPath doesn't exist"))
+        case (_, true)     => F.raiseError(new FileAlreadyExistsException(s"File $withExt already exists"))
+        case (true, false) =>
+          val absOldPath = pathOf(oldPath)
+          val absNewPath = pathOf(withExt)
+          F.delay {
+            Files.move(absOldPath, absNewPath)
+          }.as(withExt)
+      }
+    }
+  }
+
+  // TODO: should probably have a "trash" or something instead – a way of recovering a file from accidental deletion?
```

Comment:
+1, though I think this could be handled by whatever we end up doing for proper versioning so it seems fine to wait. The prompt says "Permanently delete", after all...